### ISSUE TITLE
How to add NNShark support to custom plugins?

### ISSUE THREAD

**Issue by espiriki** [nnstreamer/nnstreamer#3570]
I'm writing custom plugins for a custom video pipeline, and I would like to leverage NNShark tracing capabilities

Right now, when I run a pipeline with my custom plugins, they show zero on the timings on that GUI that's draw on the terminal.

I used gst_tracer_record_log in some specific parts of the code to log some timings, I can see the logs on the command line only. Should I use another API for that?

I digged through the source code of some plugins but couldn't find anything that hints on how this is done

Thanks!

**Comment by myungjoo** [comment]
I don't think we have those who understand the internals of gstshark/nnshark here.
And, as you have mentioned, we have done nothing additionally to have it tracked by gstshark/nnshark in the nnstreamer code.

Anyway, @wooksong @again4you @jaeyun-jung , do you have any ideas on this?

@espiriki : can you elaborate what's this "custom plugin"? Is it a tensor-filter custom plugin? or a new tensor-fitler subplugin? Can you share a minimum reproducing code?

**Comment by espiriki** [comment]
This is closed source code. But is basically one plugin that takes a frame, runs inference on it, and the output is of the type other/tensors.

Then there's another plugin that takes those tensors and runs a post processing function to get the bounding boxes (on a detection case), and then spits out those bounding boxes data to another plugin that will use cairooverlay to draw them on the frame buffer

In the future we'll update those plugins to be compatible with other NNStreamer plugins, and we want to compare our timing with those from NNStreamer as well, that's why we wanted to make those plugins compatible with NNShark

**Comment by myungjoo** [comment]
Are they GStreamer elements, inheriting GST_* classes? or NNStreamer subplugins?
Because we have done nothing in NNStreamer-side to make it compatible with GSTShark or NNShark, we'll need something to look at to see (you don't need to expose your closed source, you can show us a simple and minimized, but issue-reproducing code) what might be wrong.

Have you tested with the original GSTShark, instead of NNShark?

**Comment by espiriki** [comment]
They are gstreamer elements inheriting GST_* classes, I'll write a dummy plugin that does nothing and share the code here
I haven't tested with the original GSTShark yet, might be a good test

**Comment by espiriki** [comment]
Here it is:

https://github.com/espiriki/dummy_gst_plugin

I've also added a screen shot of the timing being zeroed out in that repo

You can test with a pipeline like this:

gst-launch-1.0 v4l2src device=/dev/videoX ! video/x-raw,width=640,height=480 ! videoconvert ! dummy ! autovideosink

It will draw a yellow square on the screen
I've added a videoconvert in the pipeline just so a valid time appears on the GUI

Let me know if you can reproduce the code

**Comment by anyj0527** [comment]
@espiriki Sorry for this late response.

With my humble glimpse,

nnshark's liveprofiler use "pad-push" hook to update infos. And I think it would work for most elements (cause they would call `gst_pad_push` in anyway).

And check if your element has single src pad and single sink pad.
https://github.com/nnstreamer/nnshark/blob/master/plugins/tracers/gstliveunit.c#L6

Anyway, I can see the live info of this example:
https://github.com/nnstreamer/nnstreamer-example/blob/main/native/example_pose_estimation_tflite/nnstreamer_example_pose_estimation_tflite.cc#L563

It shows info of cairooverlay element (the name is `tensor_res`).

![nnshark_screenshot](https://user-images.githubusercontent.com/5564297/143174700-a31e61ba-109d-4a9f-a4c7-6f025f65f5a8.png)

Please check.
